Porting Guides 2.4

gundalow opened the pull request:
##### SUMMARY

Improve the user facing documentation to inform people what changes they may need to make when upgrading Ansible.

Porting guide for 2.4 only

**Note:** 2.0 and 2.3 porting guides have been moved to https://github.com/ansible/ansible/pull/27632

##### ISSUE TYPE
 - Docs Pull Request

##### COMPONENT NAME
docs

##### ANSIBLE VERSION
```
2.3.0
2.4.0
```

gundalow committed:
Porting Guides for Ansible Versions

Improve the user facing documentation.
Tell people what changes they need to make

calfonso commented on docs/docsite/rst/porting_guide_2.3.rst:
```diff
@@ -0,0 +1,190 @@
+
+Major changes in popular modules are detailed here
+
+Modules removed
```
Is this section needed for parity with previous versions of this document? If not, is it needed if it doesn't provide any info or than no info?

gundalow replied:
I wanted to keep the order roughly the same on the different pages, hence the headings.

However since there isn't anything here I've commented it out.

thaumos commented:
I'd like to see this Guide in the sidebar, or somewhere in TOC.  Where did you think this doc would live in the tree?

dharmabumstead commented on docs/docsite/rst/porting_guide_2.0.rst:
```diff
@@ -1,9 +1,15 @@
 
+This section discusses the changes between Ansible 1.x and 2.0.
+
+It is intended to assist in updating your playbooks, plugins, etc so they will work with this version of Ansible.
```
Please avoid using 'etc'.

gundalow committed:
More porting guides

gundalow committed:
Link porting guides from TOC

gundalow committed:
Link porting guides from CHANGELOG.md

gundalow committed:
typo in link

dharmabumstead commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,74 @@
+Use of multiple tags
+--------------------
+
+Specifying ``--tags`` (or ``--skip-tags``) multiple times on the command line currently leads to the last one overriding all the previous ones. This behavior is deprecated. In the future, if you specify --tags multiple times the tags will be merged together. From now on, using ``--tags`` multiple times on one command line will emit a deprecation warning. Setting the ``merge_multiple_cli_tags`` op ...
```
Please change to "...currently leads to the last tag specified overriding all of the previously specified tags."

gundalow replied:
To this?
> From now on, using ``--tags`` multiple times on one command line will emit a deprecation warning. Setting the ``merge_multiple_cli_tags`` option to True in the ``ansible.cfg`` file currently leads to the last tag specified overriding all of the previously specified tags.

bcoca replied:
- add in 2.4 you can use multiple `-i /path/to/inventory`

dharmabumstead commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,74 @@
+
+Specifying ``--tags`` (or ``--skip-tags``) multiple times on the command line currently leads to the last one overriding all the previous ones. This behavior is deprecated. In the future, if you specify --tags multiple times the tags will be merged together. From now on, using ``--tags`` multiple times on one command line will emit a deprecation warning. Setting the ``merge_multiple_cli_tags`` op ...
+
+In 2.4, the default will be to merge and you can enable the old overwriting behavior via the config option.
```
Please revise to: "In 2.4, the default will be to merge the tags. You can enable the old overwriting behavior via the config option.

dharmabumstead commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,74 @@
+Deprecation notices
+-------------------
+
+Notice had been given that the following modules will be removed in Ansible 2.5, we suggest updating your Playbooks at your earliest convenience.
```
Please revise to: "The following modules will be removed in Ansible 2.5. Please update your playbooks accordingly."

dharmabumstead commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,74 @@
+Notice had been given that the following modules will be removed in Ansible 2.5, we suggest updating your Playbooks at your earliest convenience.
+
+* :ref:`fixme <fixme>`
+oteworthy module changes
```
"Noteworthy"

dharmabumstead commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,74 @@
+Networking
+==========
+
+There have been a number of changes to number of changes to how Networking Modules operate.
```
Please remove duplicate "number of changes to"

dharmabumstead commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,74 @@
+
+There have been a number of changes to number of changes to how Networking Modules operate.
+
+Playbooks should still use ``connection: local``
```
Add a period.

dharmabumstead commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,74 @@
+
+The following changes apply to:
+
+* TBD List modules that have been ported to new framework in 2.4
```
D the TBD

dharmabumstead commented on docs/docsite/rst/porting_guides.rst:
```diff
@@ -0,0 +1,12 @@
+Ansible Porting Guides
+**********************
+
+This section links to the various porting guides that help you upgrade your playbooks and modules from one version of Ansible to the next. Although this isn't a complete list it should assist you. I you believe any extra information would be useful in these pages you can edit by clicking ` Edit on GitHub` on the top right, or raising an issue
```
"This section lists porting guides that can help you upgrade your playbooks and modules from one version of Ansible to the next. Please note that this is not a complete list. If you believe any extra information would be useful in these pages, you can edit by clicking ` Edit on GitHub` on the top right, or raising an issue."

dharmabumstead commented on docs/docsite/rst/porting_guide_2.0.rst:
```diff
@@ -1,9 +1,19 @@
+
+This section discusses the changes between Ansible 1.x and 2.0.
+
+It is intended to assist in updating your playbooks, plugins, etc so they will work with this version of Ansible.
```
Please avoid using "etc". Suggest changing to either simply "playbooks and plugins" or maybe "your Ansible infrastructure".

dharmabumstead commented on docs/docsite/rst/porting_guide_2.0.rst:
```diff
@@ -151,7 +161,7 @@ Should now be::
-~~~~~~~~~~~~~
+-------------
 
 Here are some corner cases encountered when updating, these are mostly caused by the more stringent parser validation and the capture of errors that were previouslly ignored.
```
Period after the word "updating", then a new sentence.

dharmabumstead commented on docs/docsite/rst/porting_guide_2.0.rst:
```diff
@@ -260,13 +270,13 @@ populates the callback with them.  Here's a short snippet that shows you how:
---------------
+==============
 
 In specific cases you may want a plugin that supports both ansible-1.9.x *and* ansible-2.0. Much like porting plugins from v1 to v2, you need to understand how plugins work in each version and support both requirements. It may mean playing tricks on Ansible.
```
"playing tricks on Ansible" is ambiguous at best. Please revise and be specific.

gundalow replied:
Good spot. Not my words. Going to delete this sentence as it doesn't add any value

dharmabumstead commented on docs/docsite/rst/porting_guide_2.0.rst:
```diff
@@ -288,7 +298,8 @@ You may find the following tips useful:
-~~~~~~~~~~~~~~
+--------------
+
 As a simple example we are going to make a hybrid ``fileglob`` lookup plugin.  The ``fileglob`` lookup plugin is pretty simple to understand
```
Please remove "The ``fileglob`` lookup plugin is pretty simple to understand".

dharmabumstead commented on docs/docsite/rst/porting_guide_2.3.rst:
```diff
@@ -0,0 +1,193 @@
+Ansible 2.3 Porting Guide
+*************************
+
+This section discusses the changes to how you use Ansible between Ansible 2.2 and Ansible 2.3.
```
"This section discusses the behavioral changes between Ansible 2.2 and Ansible 2.3."

dharmabumstead commented on docs/docsite/rst/porting_guide_2.3.rst:
```diff
@@ -0,0 +1,193 @@
+
+This section discusses the changes to how you use Ansible between Ansible 2.2 and Ansible 2.3.
+
+It is intended to assist in updating your playbooks, plugins, etc so they will work with this version of Ansible.
```
Please avoid using "etc". Suggest changing to either simply "playbooks and plugins" or maybe "your Ansible infrastructure".

dharmabumstead commented on docs/docsite/rst/porting_guide_2.3.rst:
```diff
@@ -0,0 +1,193 @@
+It is intended to assist in updating your playbooks, plugins, etc so they will work with this version of Ansible.
+
+
+We suggest you read this page along with https://github.com/ansible/ansible/blob/devel/CHANGELOG.md#23-ramble-on---2017-04-12 before upgrading Ansible to understand what, if any updates you will need to make.
```
"to understand what (if any) updates you will need to make."  Although removing the "(if any)" altogether might be better.

dharmabumstead commented on docs/docsite/rst/porting_guide_2.3.rst:
```diff
@@ -0,0 +1,193 @@
+Use of multiple tags
+--------------------
+
+Specifying ``--tags`` (or ``--skip-tags``) multiple times on the command line currently leads to the last one overriding all the previous ones. This behaviour is deprecated. In the future, if you specify --tags multiple times the tags will be merged together. From now on, using ``--tags`` multiple times on one command line will emit a deprecation warning. Setting the ``merge_multiple_cli_tags`` o ...
```
Please change to "...currently leads to the last tag specified overriding all of the previously specified tags."

gundalow replied:
Is this what you suggested?
Setting the ``merge_multiple_cli_tags`` option to True in the ``ansible.cfg`` file currently leads to the last tag specified overriding all of the previously specified tags.

dharmabumstead commented on docs/docsite/rst/porting_guide_2.3.rst:
```diff
@@ -0,0 +1,193 @@
+
+Specifying ``--tags`` (or ``--skip-tags``) multiple times on the command line currently leads to the last one overriding all the previous ones. This behaviour is deprecated. In the future, if you specify --tags multiple times the tags will be merged together. From now on, using ``--tags`` multiple times on one command line will emit a deprecation warning. Setting the ``merge_multiple_cli_tags`` o ...
+
+In 2.4, the default will be to merge and you can enable the old overwriting behaviour via the config option.
```
Please revise to: "In 2.4, the default will be to merge the tags. You can enable the old overwriting behavior via the config option."

dharmabumstead commented on docs/docsite/rst/porting_guide_2.3.rst:
```diff
@@ -0,0 +1,193 @@
+Other caveats
+-------------
+
+Here are some corner cases encountered when updating, these are mostly caused by the more stringent parser validation and the capture of errors that were previously ignored.
```
Please revise to "Here are some rare cases that might be encountered when updating. These are mostly caused by the more stringent parser validation and the capture of errors that were previously ignored."

dharmabumstead commented on docs/docsite/rst/porting_guide_2.3.rst:
```diff
@@ -0,0 +1,193 @@
+Deprecation notices
+-------------------
+
+Notice had been given that the following modules will be removed in Ansible 2.5, we suggest updating your Playbooks at your earliest convenience.
```
Please revise to: "The following modules will be removed in Ansible 2.5. Please update your playbooks accordingly."

dharmabumstead commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,74 @@
+
+This section discusses the changes to how you use Ansible between Ansible 2.2 and Ansible 2.3.
+
+It is intended to assist in updating your playbooks, plugins, etc so they will work with this version of Ansible.
```
Please avoid using "etc". Suggest changing to either simply "playbooks and plugins" or maybe "your Ansible infrastructure".

dharmabumstead commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,74 @@
+It is intended to assist in updating your playbooks, plugins, etc so they will work with this version of Ansible.
+
+
+We suggest you read this page along with https://github.com/ansible/ansible/blob/devel/CHANGELOG.md#24-dancing-days---active-development before upgrading Ansible to understand what, if any updates you will need to make.
```
Suggest changing to "...to understand what (if any) updates you will need to make." Although removing the "(if any)" altogether might be better.

gundalow replied:
How about `to understand what updates you may need to make.` to make it clear that it's possible that no updates will be required.

dharmabumstead commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,74 @@
+
+We suggest you read this page along with https://github.com/ansible/ansible/blob/devel/CHANGELOG.md#24-dancing-days---active-development before upgrading Ansible to understand what, if any updates you will need to make.
+
+This document is part of a collection on porting, the rest can be found at :ref:`porting_guides <Porting Guides>`.
```
perod after "porting", then new sentence "The complete list of porting guides can be found..."

dharmabumstead reviewed (changes requested):
See edits. Thanks.

gundalow committed:
Review feedback

gundalow committed:
Better CHANGELOG links

gundalow committed:
Better links

gundalow committed:
code blocks

gundalow committed:
Minor changes

gundalow committed:
Correct links, comment out network section in 2.4

gundalow commented:
ready_for_review

I believe all comments have been addressed and would welcome a final review.

Thanks in advance

twouters commented on docs/docsite/rst/porting_guide_2.0.rst:
```diff
@@ -1,14 +1,26 @@
---------
+========
+
+This section duscusses any changes you may need to make to your playbooks.
```
discusses? :-)

gundalow replied:
My fault for using a new RST editor that doesn't highlight typos, thanks.

gundalow committed:
Typos

bcoca commented on docs/docsite/rst/porting_guide_2.3.rst:
```diff
@@ -0,0 +1,194 @@
+
+Here are some rare cases that might be encountered when updating. These are mostly caused by the more stringent parser validation and the capture of errors that were previously ignored.
+
+* The version and release facts for OpenBSD hosts were reversed. This has been changed so that version has the numeric portion and release has the name of the release.
```
I would add actual example with full varname, i would move this to playbooks section above as these vars are normally used in plays

gundalow replied:
2.3 guide moved to other PR and has been merged. This issue has been addressed.

bcoca commented on docs/docsite/rst/porting_guide_2.3.rst:
```diff
@@ -0,0 +1,194 @@
+Modules
+=======
+
+Major changes in popular modules are detailed here.
```
`No major changes in this version` should be the default for each section that is empty

bcoca commented on docs/docsite/rst/porting_guide_2.3.rst:
```diff
@@ -0,0 +1,194 @@
+
+oVirt/RHV
+^^^^^^^^^
+
```
I would remove oVirt/RHV this is 'new features' in module, unlike aws lambda and mount which change module behaviour and would require users to  reexamine existing playbooks.

bcoca commented on docs/docsite/rst/porting_guide_2.3.rst:
```diff
@@ -0,0 +1,194 @@
+.. code-block:: yaml
+
+   - name: Gather facts
+     - eos_facts:
```
incorrect indentation and extra `-`

bcoca commented on docs/docsite/rst/porting_guide_2.3.rst:
```diff
@@ -0,0 +1,194 @@
+
+Playbook
+========
+
```
Add the 'restructued async to work with action plugins' and the fact that this restores the usability of service + async

bcoca replied:
Add 'name' new to blocks .. so people can remove ugly` # commetn` on top of blocks

gundalow replied:
2.3 guide moved to other PR and has been merged. This issue has been addressed.

bcoca commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,81 @@
+
+Playbook
+========
+
```
missing:
- import_/include_ split
- ansible_facts namespacing
- inventory plugins: starting migration from hardcoded inventory + inventory scripts. scripts will still work via script plugin but efforts will now concentrate on plugins

bcoca commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,81 @@
+      
+   Plugins
+   =======
+   
```
vars plugins have changed ... will fill in details on porting

bcoca commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,81 @@
+   
+   Porting custom scripts
+   ======================
+
```
inventory scripts are becoming obsolete, in favor of inventory plugins

bcoca commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,81 @@
+   Deprecation notices
+   -------------------
+   
+   The following modules will be removed in Ansible 2.6. Please update update your    playbooks accordingly.
```
deprecation cycle is +4 now so it would be 2.8

gundalow commented:
Add
Python 2.4 and 2.5 support discontinuation

Ansible will not support Python 2.4 nor 2.5 on the target hosts anymore. Going forward, Python 2.6+ will be required on targets, as already is the case on the controller.

gundalow commented:
Update the link in https://github.com/ansible/ansible/pull/27178/files

gundalow committed:
Revert non 2.4 changes

gundalow committed:
Add placeholders for other items

gundalow committed:
multiple inventory

calfonso committed:
Update porting_guide_2.4.rst

calfonso committed:
Update porting_guide_2.4.rst

calfonso committed:
Update porting_guide_2.4.rst

calfonso committed:
Update porting_guide_2.4.rst

dharmabumstead reviewed (dismissed):
Looks good to me, except for the remaining 'FIXME' tags. Please address those before merging. Thanks @gundalow!

nitzmahone commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,175 @@
+
+The following modules will be removed in Ansible 2.8. Please update update your playbooks accordingly.
+
+* :ref:`fixme <fixme>`
```
Don't know what format we're going for with this, but:
The `azure` module has been replaced by `azure_rm_virtualmachine`, which uses the new Resource Manager SDK.

gundalow replied:
Added in https://github.com/ansible/ansible/pull/24539/commits/33ab095f16a07d500f3509a6d87bef1a8b279691#diff-d11efc4cd1f939e08f3bd127476edb26R109

nitzmahone committed:
add win_command/win_shell escaping notes

jborean93 committed:
Added some windows porting info

dharmabumstead committed:
Edit for grammar

gundalow committed:
Correct links, delete Net Platform Agnostic

gundalow committed:
2.5 link

ganeshrn commented on docs/docsite/rst/porting_guide_2.4.rst:
```diff
@@ -0,0 +1,151 @@
+   [WARNING]: deprecation message 3
+
+
+**NEW** In Ansible 2.4:
```
Persistent config vars `connect_retries` and `connect_interval ` available in 2.3 is deprecated.
From `2.4` onward to set socket to connect timeout need to set variable `connect_retry_timeout` 

New timer added to control the command response timeout `command_timeout` in 2.4
Previously in 2.3 it was controlled using global timeout ie. `timeout` variable under `default`.

gundalow committed:
Network changes in 2.4

ganeshrn reviewed (approved):
Looks good for network section

abadger committed:
facts_namespacing was pushed back to 2.5

Due to lack of time we pushed this back to 2.5 so that the facts in the namespace could be shortened.

gundalow committed:
Python version

gundalow committed:
Merge branch 'docs-porting-guides' of github.com:gundalow/ansible into docs-porting-guides

gundalow committed:
Remove FIXMEs - move to seperate PR

gundalow commented:
As we are very close to 2.4.0FINAL and in discussion with @abadger I've removed the FIXME sections and I'll create a follow up PR which @bcoca and others can update.

This will allow us to regenerate `docs.ansible.com/ansible` along side `2.4.0rc4`

gundalow commented:
`FIXMEs` have been moved to https://github.com/ansible/ansible/pull/30456

abadger commented:
hcerry picked for 2.4.0rc4
